modules/mailer/mail_issue.go

```diff
@@ -118,39 +118,41 @@ func mailIssueCommentToParticipants(e Engine, issue *Issue, doer *User, content
 
 // MailParticipants sends new issue thread created emails to repository watchers
 // and mentioned people.
-func (issue *Issue) MailParticipants(doer *User, opType ActionType) (err error) {
-	return issue.mailParticipants(x, doer, opType)
+func MailParticipants(issue *models.Issue, doer *models.User, opType models.ActionType) (err error) {
+	return models.WithEngine(func(e models.Engine) error {
+		return mailParticipants(issue, e, doer, opType)
+	})
 }
 
-func (issue *Issue) mailParticipants(e Engine, doer *User, opType ActionType) (err error) {
+func mailParticipants(issue *models.Issue, e models.Engine, doer *models.User, opType models.ActionType) (err error) {
 	mentions := markup.FindAllMentions(issue.Content)
 
-	if err = UpdateIssueMentions(e, issue.ID, mentions); err != nil {
+	if err = models.UpdateIssueMentions(e, issue.ID, mentions); err != nil {
 		return fmt.Errorf("UpdateIssueMentions [%d]: %v", issue.ID, err)
 	}
 
 	if len(issue.Content) > 0 {
-		if err = mailIssueCommentToParticipants(e, issue, doer, issue.Content, nil, mentions); err != nil {
+		if err = mailIssueCommentToParticipants(issue, doer, issue.Content, nil, mentions); err != nil {
```

Comment:
You've dropped the engine from this function which basically means that passing in the engine in models.UpdateIssueMentions sort of pointless - you can't do this all in one transaction anyway. So at this point we could just drop the engine

Reply:
Yes, it should not in a transaction. Sending mail failed should only warn user or log.